src/component/modal/editRange/EditRangeModal.jsx

```diff
@@ -69,48 +62,135 @@ const styles = css`
   }
 `;
 
-const EditRangeModal = ({ onSave, onClose, onZoom, range }) => {
-  const dispatch = useDispatch();
-
+const EditRangeModal = ({
+  onSaveHandler,
+  onCloseHandler,
+  onZoomHandler,
+  range,
+}) => {
   const handleOnZoom = useCallback(() => {
-    onZoom(range);
-  }, [onZoom, range]);
+    onZoomHandler(range);
+  }, [onZoomHandler, range]);
 
   useEffect(() => {
     handleOnZoom();
-  }, [dispatch, handleOnZoom, range]);
+  }, [handleOnZoom]);
 
   const handleOnClose = useCallback(() => {
-    onClose();
-  }, [onClose]);
+    onCloseHandler();
+  }, [onCloseHandler]);
+
+  const getCouplings = useCallback(
+    (couplings) =>
+      couplings
+        .filter((coupling) => coupling.coupling !== '')
+        .map((coupling) => {
+          return {
+            ...coupling,
+            multiplicity: translateMultiplet(coupling.multiplicity),
+          };
+        }),
+    [],
+  );
+
+  const getSignals = useCallback(
+    (signals) => {
+      return signals.map((_signal) => {
+        return {
+          ..._signal,
+          multiplicity: _signal.j
+            .map((_coupling) => translateMultiplet(_coupling.multiplicity))
+            .join(''),
+          j: getCouplings(_signal.j),
+        };
+      });
+    },
+    [getCouplings],
+  );
 
   const handleOnSave = useCallback(
     async (formValues) => {
-      range.signal = formValues.signals.slice();
-      await onSave(range);
+      const _range = lodash.cloneDeep(range);
+      _range.signal = getSignals(formValues.signals);
+      await onSaveHandler(_range);
       handleOnClose();
     },
-    [handleOnClose, onSave, range],
+    [getSignals, handleOnClose, onSaveHandler, range],
   );
 
+  const initialStateSignals = useMemo(() => {
+    return range.signal.map((_signal) => {
+      // counter within j array to access to right j values
+      let counterJ = 0;
+      const couplings = [];
+      let coupling;
+      _signal.multiplicity.split('').forEach((_multiplicity) => {
+        if (hasCouplingConstant(_multiplicity)) {
+          coupling = { ..._signal.j[counterJ] };
+          counterJ++;
+        } else {
+          coupling = { multiplicity: _multiplicity, coupling: '' };
+        }
+        coupling.multiplicity = translateMultiplet(coupling.multiplicity);
+        couplings.push(coupling);
+      });
+      return { ..._signal, j: couplings };
+    });
+  }, [range.signal]);
+
+  const isSaveButtonDisabled = useCallback((errors) => {
+    if (Object.keys(errors).length > 0) {
+      if (
+        // ignore non-relevant newSignalDelta field
+        Object.keys(errors).length === 1 &&
+        Object.keys(errors)[0] === 'newSignalDelta'
```

Comment:
Why to call object.keys() three times, you can call it once above and assign the result to variable and then use it inside the the if statement